### ISSUE TITLE
Clang warning insists on changing format specifier from `%u` to `%u`

### ISSUE THREAD

**Issue by johnstiles-google** [llvm/llvm-project#63867]
Test case: https://godbolt.org/z/aW3Eax35r

We have code that uses `printf` to output an `enum class : unsigned int` with a `%u` format specifier:
```
enum class E : unsigned int {
    A, B, C
};

void test(E e) {
    printf("%u", e);
}
```

 Previously (clang 16.0.0) this was benign and compiled without incident. On clang-trunk in -Wall we get an incorrect warning:
 
```
<source>:8:18: warning: format specifies type 'unsigned int' but the argument has type 'E' [-Wformat]
    8 |     printf("%u", e);
      |             ~~   ^
      |             %u
```

**Comment by shafik** [comment]
The diagnostic should file, gcc also diagnoses this case: https://godbolt.org/z/Gfraczh8h

The fix-it is confused, I don't think we should be offering a fix-it the correct solution is to `static_cast` the enum to unsigned.

**Comment by johnstiles-google** [comment]
I'm not convinced that GCC has got it right.

https://en.cppreference.com/w/cpp/language/variadic_arguments

"Only arithmetic, enumeration, pointer, pointer to member, and class type arguments (after conversion) are allowed." If enumerated types are allowed, and the exact base type is known, I should not need a cast.

**Comment by efriedma-quic** [comment]
Relevant standard bits:

C++ [expr.call]: If the argument has integral or enumeration type that is subject to the integral promotions, or a floating-point type that is subject to the floating-point promotion, the value of the argument is converted to the promoted type before the call. These promotions are referred to as the default argument promotions.

C++ [conv.prom]: A prvalue of an unscoped enumeration type whose underlying type is fixed ([dcl.enum]) can be converted to a prvalue of its underlying type.

C "The va_arg macro": If there is no actual next argument, or if type is not compatible with the type of the actual next argument (as promoted according to the default argument promotions), the behavior is undefined, except for the following cases: one type is a signed integer type, the other type is the corresponding unsigned integer
type, and the value is representable in both types; [...]

Unscoped enums get promoted to the underlying type, which then is compatible with a printf "int" or "unsigned int" specifier.  None of this text applies to scoped enums.  So by default, you get the most strict rule: the type passed to va_arg has to exactly match the type of the enum.

[expr.call] also says "Passing a potentially-evaluated argument of a scoped enumeration type [...] is conditionally-supported with implementation-defined semantics", but I think that's only relevant if you call va_arg with the exact enum type, which printf doesn't.

Granted, the C++ standard text is a little weird is a little weird because [cstdarg.syn] effectively imports the C rules for a "compatible type", but that isn't really relevant here.

**Comment by johnstiles-google** [comment]
OK, I'll concede to the standard on scoped-enums, so I think the only missing thing here is that the warning text is misleading.

**Comment by shafik** [comment]
CC @abrachet

**Comment by abrachet** [comment]
We've just fixed this in 563a23c82
